### ISSUE TITLE
HttpAuth: User auth cache needs to be deleted when user is deleted 

### ISSUE THREAD

**Issue by sfitts** [grafana/grafana#21340]
**What happened**:

When deleting a user, any corresponding entry in the auth cache isn't also invalidated.

**What you expected to happen**:

For the deleted user's auth cache entry to be invalidated.

**How to reproduce it (as minimally and precisely as possible)**:

## @dpavlos' Account
According to @dpavlos, it's a proxy cache related issue and can be reproduced as follows:

1. start from an empty db
2. create the user
3. set isGrafanaAdmin to true
4. authenticate once using the X-WEBAUTH-USER: <username>. Everything works as expected.
5. delete the user
6. repeat steps 2,3,4

When you delete the user, the proxy key that was generated and saved in cache_data table in step 4 is not invalidated properly. Therefore, when you recreate the user with the same username and then try to authenticate, the Login function checks if there are any proxy keys associated with that username, https://github.com/grafana/grafana/blob/35679d4689fb08ac4d64a857d616bc7ced5cd22f/pkg/middleware/auth_proxy/auth_proxy.go#L201, finds the old one (since it hasn't expired yet) and returns the ID of the already deleted user. Hence, a "User not found" message is returned when you call the /stats endpoint for the second time.

## @sfitts' Outdated Account

Issue the following curl commands:

```
$ curl -X POST -H "X-WEBAUTH-USER: admin" -H "Content-Type: application/json" -d'{"login": "vantiqSystem", "password": "unused"}' http://localhost:3000/api/admin/users
{
  "id": 1480,
  "message": "User created"
}

$ curl -X PUT -H "X-WEBAUTH-USER: admin" -H "Content-Type: application/json" -d'{"isGrafanaAdmin": true}' http://localhost:3000/api/admin/users/1480/permissions
{
  "message": "User permissions updated"
}

$ curl -H "X-WEBAUTH-USER: admin" http://localhost:3000/api/admin/stats
{
  "orgs": 14,
  "dashboards": 97,
  "snapshots": 0,
  "tags": 0,
  "datasources": 13,
  "playlists": 0,
  "stars": 0,
  "alerts": 0,
  "users": 72,
  "admins": 14,
  "editors": 13,
  "viewers": 14,
  "activeUsers": 71,
  "activeAdmins": 14,
  "activeEditors": 13,
  "activeViewers": 13,
  "activeSessions": 0
}

$ curl -vis -H "X-WEBAUTH-USER: vantiqSystem" http://localhost:3000/api/admin/stats

*   Trying ::1:3000...
* TCP_NODELAY set
* Connected to localhost (::1) port 3000 (#0)
> GET /api/admin/stats HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/7.65.3
> Accept: */*
> X-WEBAUTH-USER: vantiqSystem
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 407 Proxy Authentication Required
HTTP/1.1 407 Proxy Authentication Required
< Cache-Control: no-cache
Cache-Control: no-cache
< Content-Type: text/html; charset=UTF-8
Content-Type: text/html; charset=UTF-8
< Expires: -1
Expires: -1
< Pragma: no-cache
Pragma: no-cache
< X-Frame-Options: deny
X-Frame-Options: deny
< Date: Fri, 03 Jan 2020 21:10:11 GMT
Date: Fri, 03 Jan 2020 21:10:11 GMT
< Content-Length: 1727
Content-Length: 1727

<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
    <meta name="viewport" content="width=device-width">
    <meta name="theme-color" content="#000">

    <title>Grafana - Error</title>

    <base href="/grafana/" />

    <link rel="stylesheet" href="public/build/grafana.dark.d9f582b9b538bf193c78.css">

    <link rel="icon" type="image/png" href="public/img/fav32.png">
    <link rel="mask-icon" href="public/img/grafana_mask_icon.svg" color="#F05A28">

  </head>

  <body class="theme-dark">
    <div class="main-view">
      <div class="page-container">
        <div class="page-header">
          <div class="page-header__inner">
            <span class="page-header__logo">
              <i class="page-header__icon fa fa-frown-o"></i>
            </span>
            <div class="page-header__info-block">
              <h1 class="page-header__title">
                <a class="text-link" href="login">Grafana</a><span> / Server Error</span><span></span>
              </h1>
              <div class="page-header__sub-title">Sadly something went wrong</div>
            </div>
          </div>
        </div>
      </div>
      <div class="page-container page-body ng-scope" style="padding: 2rem">
        <div class="alert">
          <div class="alert-icon"><i class="fa fa-exclamation-triangle"></i></div>
          <div class="alert-body">
            <div class="alert-title">User not found</div>
          </div>
        </div>
        <br />

        <div style="padding: 2rem 0 0">
          <p>Check the Grafana server logs for the detailed error message.</p>
        </div>
      </div>
    </div>
  </body>
</html>
```

Contents of log for failed request:

```
t=2020-01-03T21:02:37+0000 lvl=info msg="Request Completed" logger=context userId=1 orgId=1 uname=admin method=GET path=/api/admin/stats status=200 remote_addr=172.21.0.1 time_ms=4 size=312 referer=
t=2020-01-03T21:02:48+0000 lvl=eror msg="Failed to get signed user" logger=auth.proxy username= message="User not found" error=nil
```

**Anything else we need to know?**:

**Environment**:
- Grafana version: 6.5.2
- OS Grafana is installed on: tested on both Windows and Linux (both via Docker).
- Others:  Grafana configured as follows:

```
instance_name = vantiq_grafana
app_mode = development

[server]
root_url = /grafana
router_logging = true

[database]
type = mysql
host = vantiq_mysql:3306

[users]
allow_sign_up = false
auto_assign_org = true
allow_org_create = false

[auth]
disable_login_form = true

[auth.basic]
enabled = false

[auth.proxy]
enabled = true
header_name = X-WEBAUTH-USER
header_property = username
auto_sign_up = false
enable_login_token = true
```

**Comment by sfitts** [comment]
This is blocking us from upgrading to Grafana 6.x since currently we cannot provision users.

**Comment by sfitts** [comment]
The same issue arises when you create a non-admin user which you then associate with an org.  It used to be possible to provision users in this manner and they would end up with the permissions to the org as assigned.  Now they don't work at all.  

Another interesting fact is that users that existed prior to the upgrade from 5.4 to 6.5 continue to work just fine.  This only impacts newly created users for some reason.  One thing I've noticed is that migrated users have some extra information in their user record.  Here is a migrated user:

```
{
    "authLabels": [
        "OAuth"
    ],
    "createdAt": "2020-01-03T20:41:43Z",
    "email": "local_MyNamespace_90922_editor",
    "id": 1466,
    "isDisabled": false,
    "isExternal": true,
    "isGrafanaAdmin": false,
    "login": "local_MyNamespace_90922_editor",
    "name": "",
    "orgId": 427,
    "theme": "",
    "updatedAt": "2020-01-03T20:41:43Z"
}
```

And here is one that is created directly on 6..5.2:

```
{
    "authLabels": [],
    "createdAt": "2020-01-03T20:53:22Z",
    "email": "vantiqSystem",
    "id": 1480,
    "isDisabled": false,
    "isExternal": false,
    "isGrafanaAdmin": true,
    "login": "vantiqSystem",
    "name": "",
    "orgId": 1,
    "theme": "",
    "updatedAt": "2020-01-03T20:53:22Z"
}
```

Note that the migrated user has the "authLabels" while the new one does not.  I've tried adding that to my newly created users via PUT, but that change is ignored.

**Comment by torkelo** [comment]
Unable to replicate this. 

Enable auth proxy like in your settings. and executed your commands:

```
❯ curl -X GET -H "X-WEBAUTH-USER: vantiqSystem" http://localhost:3000/api/admin/stats
{
  "orgs": 2,
  "dashboards": 251,
  "snapshots": 3,
  "tags": 39,
  "datasources": 45,
  "playlists": 3,
  "stars": 6,
  "alerts": 8,
  "users": 8,
  "admins": 2,
  "editors": 2,
  "viewers": 4,
  "activeUsers": 2,
  "activeAdmins": 1,
  "activeEditors": 0,
  "activeViewers": 1,
  "activeSessions": 1
}
```

**Comment by sfitts** [comment]
Did you have `enable_login_token` set to `true`?  This only reproduces for me with that setting (if it is set to `false` then things work as expected).  We need that set so that Grafana will generate the session cookie when we visit `/login` (otherwise we need to add the proxy auth header to every request made by the browser, which is a non-starter for us.  Related to https://github.com/grafana/grafana/issues/17316).

**Comment by sfitts** [comment]
It is even trickier than that.  Turns out that even with `enable_login_token` set to `true` from an empty database it takes 2 passes to reproduce.  So, starting from an empty database, using the settings shown above, here's what I see:

```
$ curl -X POST -H "X-WEBAUTH-USER: admin" -H "Content-Type: application/json" -d'{"login": "vantiqS
ystem", "password": "unused"}' http://localhost:3000/api/admin/users
{
  "id": 2,
  "message": "User created"
}
$ curl -X PUT -H "X-WEBAUTH-USER: admin" -H "Content-Type: application/json" -d'{"isGrafanaAdmin":
true}' http://localhost:3000/api/admin/users/2/permissions
{
  "message": "User permissions updated"
}
$ curl -H "X-WEBAUTH-USER: vantiqSystem" http://localhost:3000/api/admin/stats                      {
  "orgs": 1,
  "dashboards": 0,
  "snapshots": 0,
  "tags": 0,
  "datasources": 0,
  "playlists": 0,
  "stars": 0,
  "alerts": 0,
  "users": 2,
  "admins": 1,
  "editors": 0,
  "viewers": 1,
  "activeUsers": 2,
  "activeAdmins": 1,
  "activeEditors": 0,
  "activeViewers": 1,
  "activeSessions": 0
}
$ curl -X DELETE -H "X-WEBAUTH-USER: admin" http://localhost:3000/api/admin/users/2
{
  "message": "User deleted"
}
$ curl -X POST -H "X-WEBAUTH-USER: admin" -H "Content-Type: application/json" -d'{"login": "vantiqS
ystem", "password": "unused"}' http://localhost:3000/api/admin/users
{
  "id": 3,
  "message": "User created"
}
$ curl -X PUT -H "X-WEBAUTH-USER: admin" -H "Content-Type: application/json" -d'{"isGrafanaAdmin":
true}' http://localhost:3000/api/admin/users/3/permissions
{
  "message": "User permissions updated"
}
$ curl -H "X-WEBAUTH-USER: vantiqSystem" http://localhost:3000/api/admin/stats
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
    <meta name="viewport" content="width=device-width">
    <meta name="theme-color" content="#000">

    <title>Grafana - Error</title>

    <base href="/grafana/" />

    <link rel="stylesheet" href="public/build/grafana.dark.d9f582b9b538bf193c78.css">

    <link rel="icon" type="image/png" href="public/img/fav32.png">
    <link rel="mask-icon" href="public/img/grafana_mask_icon.svg" color="#F05A28">

  </head>

  <body class="theme-dark">
    <div class="main-view">
      <div class="page-container">
        <div class="page-header">
          <div class="page-header__inner">
            <span class="page-header__logo">
              <i class="page-header__icon fa fa-frown-o"></i>
            </span>
            <div class="page-header__info-block">
              <h1 class="page-header__title">
                <a class="text-link" href="login">Grafana</a><span> / Server Error</span><span></span>
              </h1>
              <div class="page-header__sub-title">Sadly something went wrong</div>
            </div>
          </div>
        </div>
      </div>
      <div class="page-container page-body ng-scope" style="padding: 2rem">
        <div class="alert">
          <div class="alert-icon"><i class="fa fa-exclamation-triangle"></i></div>
          <div class="alert-body">
            <div class="alert-title">User not found</div>
          </div>
        </div>
        <br />

        <div style="padding: 2rem 0 0">
          <p>Check the Grafana server logs for the detailed error message.</p>
        </div>
      </div>
    </div>
  </body>
</html>
```

**Comment by sfitts** [comment]
Note that this result from the above sequence also seems odd:

```
$ curl -H "X-WEBAUTH-USER: vantiqSystem" http://localhost:3000/api/admin/stats                      {
  "orgs": 1,
  "dashboards": 0,
  "snapshots": 0,
  "tags": 0,
  "datasources": 0,
  "playlists": 0,
  "stars": 0,
  "alerts": 0,
  "users": 2,
  "admins": 1,
  "editors": 0,
  "viewers": 1,
  "activeUsers": 2,
  "activeAdmins": 1,
  "activeEditors": 0,
  "activeViewers": 1,
  "activeSessions": 0
}
```

Given that the `vantiqSystem` user was marked as a global admin, I would have expected the `admin` count to be 2 and not 1 (but maybe that's always been true, never really looked closely).

**Comment by cantino** [comment]
We're also encountering this issue in production. AuthProxy simply doesn't work with users that we create with the API.

**Comment by dpavlos** [comment]
This seems to be a proxy cache related issue @torkelo . I managed to reproduce the error on latest master by following the steps mentioned by @sfitts :

1) start from an empty db
2) create the user 
3) set isGrafanaAdmin to true
4) authenticate once using the `X-WEBAUTH-USER: <username>`. On this step a proxy key is generated.
5) delete the user
6) repeat steps 2,3,4

When you delete the user the proxy key that were generated and saved in `cache_data` table on step 4 is not invalidated properly, so when you recreate the user with the same username, the Login function checks if there are any proxy keys associated with that username, https://github.com/grafana/grafana/blob/35679d4689fb08ac4d64a857d616bc7ced5cd22f/pkg/middleware/auth_proxy/auth_proxy.go#L201 finds the old one (since it hasn't expired yet) and returns the ID of the already deleted user. Hence, a "User not found" message is returned when you call the /stats endpoint for the second time.

**Comment by cantino** [comment]
Agreed @dpavlos, manually deleting the contents of that table seemed to clear the cache and allow login.

**Comment by bergquist** [comment]
This is a bit more frustration to solve then I expected since the cache key is generated based on what header are used in the auth_proxy. https://github.com/grafana/grafana/blob/35679d4689fb08ac4d64a857d616bc7ced5cd22f/pkg/middleware/auth_proxy/auth_proxy.go#L155 

The two options I see are 

1) Once the user have been deleted in the database we need to rebuild that key and delete the cached value. 

2) We cache two values. The current and a reserved item where the userid is key. So when the user is deleted, we could find the first cache key by looking at the cached value for the userid and then delete both. This would require twice as much cache but I doubt that's a problem.

**Comment by digrich** [comment]
We're also encountering similar issue with Grafana 6.2.5.

**Comment by digrich** [comment]
Hi @berghauz - Do you have any plan to release the fix before 7.0? We are not able to upgrade grafana beyond 6.4.3 because of this issue.

**Comment by berghauz** [comment]
@digrich you must be mixed me up

**Comment by digrich** [comment]
Sorry @berghauz, my bad. I was requesting @bergquist.

**Comment by rejeevd** [comment]
Hi,
I am also facing the same issue. Is there any workaround? Is there any fix planned?

**Comment by aknuds1** [comment]
I'm looking into this issue, and will try to solve it :)

**Comment by aknuds1** [comment]
I could reproduce. Logging off for today though, will have to check back tomorrow!

**Comment by aknuds1** [comment]
I've tried to fix this issue in #23979, please have a look.
